hathor/indexes/utxo_index.py

```diff
@@ -0,0 +1,326 @@
+# Copyright 2022 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import Iterator, Optional
+
+from structlog import get_logger
+
+from hathor.conf import HathorSettings
+from hathor.indexes.base_index import BaseIndex
+from hathor.transaction import BaseTransaction
+from hathor.transaction.scripts import parse_address_script
+from hathor.util import peekable, sorted_merger
+
+logger = get_logger()
+settings = HathorSettings()
+
+
+@dataclass(frozen=True)
+class UtxoIndexItem:
+    token_uid: bytes
+    tx_id: bytes
+    index: int
+    address: str
+    amount: int
+    timelock: Optional[int]
+    heightlock: Optional[int]
+
+    def __repr__(self):
+        return (
+            'UtxoIndexItem('
+            f'token_uid={self.token_uid.hex()},'
+            f'tx_id={self.tx_id.hex()},'
+            f'index={self.index},'
+            f'address={self.address},'
+            f'timelock={self.timelock},'
+            f'heightlock={self.heightlock}'
+            ')'
+        )
+
+
+class UtxoIndex(BaseIndex):
+    """ Index of UTXOs by address+token
+
+    This index is currently optional and only used for an optional API.
+
+    It is expected to be called with only non-voided transactions and blocks, including the genesis. It ignores
+    token-creation transactions and data-only outputs.
+    """
+
+    def __init__(self):
+        self.log = logger.new()
+
+    # interface methods provided by the base class
+
+    def init_loop_step(self, tx: BaseTransaction) -> None:
+        tx_meta = tx.get_metadata()
+        # we don't want ANY type of voided transaction
```

Comment:
What if a transaction moves from validated to voided? Or the other way around?

Reply:
A transaction cannot change from voided -> valid or valid -> voided _during_ initialization, unless using full-verification, but full-verification does not use `init_loop_step`, so this is not affected.

Reply:
This changed slightly, and now it doesn't even need to filter-out the voided ones.

Reply:
Great!